lib/BaseWasmMainTemplatePlugin.js

```diff
@@ -0,0 +1,242 @@
+/*
+	MIT License http://www.opensource.org/licenses/mit-license.php
+	Author Tobias Koppers @sokra
+*/
+"use strict";
+
+const Template = require("./Template");
+const WebAssemblyImportDependency = require("./dependencies/WebAssemblyImportDependency");
+
+// Get all wasm modules
+function getAllWasmModules(chunk) {
+	const wasmModules = chunk.getAllAsyncChunks();
+	const array = [];
+	for (const chunk of wasmModules) {
+		for (const m of chunk.modulesIterable) {
+			if (m.type.startsWith("webassembly")) {
+				array.push(m);
+			}
+		}
+	}
+
+	return array;
+}
+
+function generateImportObject(module) {
+	const depsByRequest = new Map();
+	for (const dep of module.dependencies) {
+		if (dep instanceof WebAssemblyImportDependency) {
+			// Ignore global they will be handled later
+			if (dep.description.type === "GlobalType") {
+				continue;
+			}
+
+			const request = dep.request;
+			let array = depsByRequest.get(request);
+			if (!array) {
+				depsByRequest.set(request, (array = []));
+			}
+			const exportName = dep.name;
+			const usedName = dep.module && dep.module.isUsed(exportName);
+			array.push({
+				exportName,
+				usedName,
+				module: dep.module,
+				description: dep.description
+			});
+		}
+	}
+	const importsCode = [];
+	for (const pair of depsByRequest) {
+		const properties = [];
+		for (const data of pair[1]) {
+			let params = "";
+			let result = "void 0";
+
+			if (data.description.type === "FuncImportDescr") {
+				params = data.description.params.map(
+					(param, k) => "p" + k + param.valtype
+				);
+
+				result = `__webpack_require__(${JSON.stringify(
+					data.module.id
+				)})[${JSON.stringify(data.usedName)}](${params})`;
+			}
+
+			properties.push(
+				`\n\t\t${JSON.stringify(data.exportName)}: function(${params}) {
+                                  return ${result};
+                                }`
+			);
+		}
+		importsCode.push(
+			`\n\t${JSON.stringify(pair[0])}: {${properties.join(",")}\n\t}`
+		);
+	}
+
+	return JSON.stringify(module.id) + ": {" + importsCode.join(",") + "\n},";
+}
+
+class BaseWasmMainTemplatePlugin {
+	applyWeb(mainTemplate) {
+		const generateLoadBinaryCode = path =>
+			`fetch(${mainTemplate.requireFn}.p + ${path})`;
+
+		this.apply(mainTemplate, generateLoadBinaryCode);
+	}
+
+	applyNode(mainTemplate) {
```

Comment:
It's a bit weird to mix both targets into one file. Instead move `applyNode` to `ReadFileCompileWasmMainTemplatePlugin.apply` (`ReadFileCompileWasmMainTemplatePlugin extends BaseWasmMainTemplatePlugin` and resp. for `applyWeb`

Reply:
Instead of different `apply` methods you can also make `generateLoadBinaryCode` an abstract method.

Reply:
Ok, I can refactor to use some class inheritance or I can just move the specific code from the base to the callers? What's the best to you?

Reply:
I'm more for "composition over inheritance", I changed in https://github.com/webpack/webpack/pull/6709/commits/405fea671972a9920e1aed711758a5a1e051d5b1